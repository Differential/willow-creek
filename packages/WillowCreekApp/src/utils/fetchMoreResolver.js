import get from 'lodash/get';
import set from 'lodash/fp/set';

export default function fetchMoreResolver({
  collectionName,
  variables,
  fetchMore,
  data,
} = {}) {
  return () => {
    const pageInfoPath = `${collectionName}.pageInfo`;
    const edgePath = `${collectionName}.edges`;

    const after = get(data, `${pageInfoPath}.endCursor`);
    if (!after) return;

    fetchMore({
      variables: { ...variables, after },
      updateQuery: (previousResult, { fetchMoreResult }) => {
        // combine previous data and fetchMore data
        const newDataWithPageInfo = set(
          pageInfoPath,
          {
            ...get(previousResult, pageInfoPath, {}),
            ...get(fetchMoreResult, pageInfoPath, {}),
          },
          previousResult
        );
        const newDataWithAdditionalItems = set(
          edgePath,
          [
            ...get(previousResult, edgePath, []),
            ...get(fetchMoreResult, edgePath, []),
          ],
          newDataWithPageInfo
        );

        return newDataWithAdditionalItems;
      },
    });
  };
}
