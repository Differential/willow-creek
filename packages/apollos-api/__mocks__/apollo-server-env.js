import fetch from 'jest-fetch-mock';

import * as apolloDatasourceMocks from './apollo-datasource-mocks';
import * as rockMocks from './rock-api-mocks';

const apolloServerEnv = require.requireActual('apollo-server-env');

// eslint-disable-next-line
const Response = apolloServerEnv.Response;

const resolveWith = (data, url) =>
  Promise.resolve(
    new Response(JSON.stringify(data), {
      url,
      status: 200,
      statusText: 'OK',
      headers: new apolloServerEnv.Headers({
        'Content-Type': 'application/json',
      }),
    })
  );

fetch.mockLiveDataSourceApis = () => {
  fetch.mockImplementation((request) => {
    if (request.url.match('/api/v1/events/current')) {
      return resolveWith(apolloDatasourceMocks.liveStreamLive(), request.url);
    }
    if (request.url.match('api.scripture.api.bible')) {
      return resolveWith(apolloDatasourceMocks.Scripture(), request.url);
    }
    return Promise.reject();
  });
};

fetch.mockRockDataSourceAPI = () => {
  fetch.mockImplementation((request) => {
    let { url } = request;
    url = decodeURI(url);
    if (!url.match('https://apollosrock.newspring.cc/api')) {
      if (request.url.match('/api.scripture.api.bible/v1')) {
        return resolveWith(apolloDatasourceMocks.Scripture());
      }
      return Promise.reject();
    }

    if (url.match('api/EntityTypes')) {
      return resolveWith(
        [
          {
            Id: 201,
          },
        ],
        url
      );
    }

    if (url.match('api/InteractionChannels')) {
      return resolveWith(
        [
          {
            Id: 10,
          },
        ],
        url
      );
    }

    if (url.match('api/InteractionComponents')) {
      return resolveWith(
        [
          {
            Id: 101,
          },
        ],
        url
      );
    }

    if (url.match('api/ContentChannels/\\d')) {
      return resolveWith(rockMocks.contentChannel());
    }

    if (url.match('api/Groups/GetFamilies/\\d')) {
      return resolveWith(rockMocks.families());
    }

    if (url.match('api/ContentChannels')) {
      return resolveWith([rockMocks.contentChannel()]);
    }

    if (url.match('api/ContentChannelItems/test-case-no-attributes')) {
      const mock = rockMocks.contentItem();
      mock.AttributeValues = {};
      mock.Attributes = {};
      return resolveWith(mock);
    }

    if (url.match('api/ContentChannelItems/123')) {
      const contentItem = rockMocks.contentItem();
      contentItem.AttributeValues.Scriptures = { Value: 'John 3:16' };
      return resolveWith(contentItem);
    }

    if (url.match('api/ContentChannelItems/\\d')) {
      return resolveWith(rockMocks.contentItem());
    }

    if (url.match('api/ContentChannelItems')) {
      return resolveWith([rockMocks.contentItem()]);
    }

    if (url.match('api/ContentChannelItemAssociations')) {
      return resolveWith([rockMocks.contentChannelItemAssociation()]);
    }

    if (url.match('api/Auth/Login')) {
      const body = JSON.parse(request.body);
      const response = new Response('');
      if (body.Password === 'good') {
        response.headers.set('set-cookie', 'some cookie');
        return Promise.resolve(response);
      }
      return Promise.reject({ ...response, status: 401 });
    }

    if (url.match('api/People/GetCurrentPerson')) {
      return resolveWith(rockMocks.people());
    }

    if (url.match('api/People/\\d')) {
      return resolveWith(rockMocks.people());
    }

    if (url.match('api/People')) {
      if (request.method === 'POST') {
        const { Email } = JSON.parse(request.body);
        if (!Email) {
          const response = new Response('');
          response.status = 400;
          return Promise.reject(response);
        }
        return resolveWith('35');
      }

      return resolveWith([rockMocks.people()]);
    }

    if (url.match('api/Interactions/\\d')) {
      return resolveWith(rockMocks.interaction());
    }

    if (url.match('api/Interactions')) {
      if (request.method === 'GET') {
        return resolveWith([rockMocks.interaction()]);
      }
      if (request.method === 'POST') {
        const response = new Response('456', {
          status: 200,
          statusText: 'OK',
          headers: new apolloServerEnv.Headers({
            'Content-Type': 'application/json',
          }),
        });
        return Promise.resolve(response);
      }
    }

    if (url.match('api/InteractionSessions')) {
      if (request.method === 'GET') {
        return resolveWith(rockMocks.session());
      }
      if (request.method === 'POST') {
        const response = new Response('123', {
          status: 200,
          statusText: 'OK',
          headers: new apolloServerEnv.Headers({
            'Content-Type': 'application/json',
          }),
        });
        return Promise.resolve(response);
      }
    }

    if (url.match('api/UserLogins')) {
      if (request.method === 'POST') {
        const { UserName } = JSON.parse(request.body);
        if (!UserName) {
          const response = new Response('');
          response.status = 400;
          return Promise.reject(response);
        }
        return resolveWith({ id: 21 });
      }
      const identity = url // identity = UserName
        .split('eq')
        .pop()
        .trim(' '); // EXAMPLE URL: /api/UserLogins?$filter=UserName eq 'isaac.hardy@newspring.cc'

      if (!identity) {
        const response = new Response('');
        response.status = 400;
        return Promise.reject(response);
      }
      if (identity === `'isaac.hardy@newspring.cc'`)
        return resolveWith([rockMocks.userLogins()]);

      return resolveWith([]);
    }
    console.log(`No route matching ${url}`);
    return Promise.reject(`No route matching ${url}`);
  });
};

const apolloServerEnvMocked = {
  ...apolloServerEnv,
  Response: fetch.Response,
  fetch,
};

module.exports = apolloServerEnvMocked;
