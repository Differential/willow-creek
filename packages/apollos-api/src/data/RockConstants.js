import { RockConstants } from '@apollosproject/data-connector-rock';

class dataSource extends RockConstants.dataSource {
  async createOrFindInteractionComponent({
    componentName,
    channelId,
    entityId,
  }) {
    return this.findOrCreate({
      model: 'InteractionComponents',
      objectAttributes: {
        Name: componentName,
        InteractionChannelId: channelId, // In willow, this field is InteractionChannel instead of Channel for unknown reasons.
        EntityId: entityId,
      },
    });
  }
}
// eslint-disable-next-line import/prefer-default-export
export { dataSource };
