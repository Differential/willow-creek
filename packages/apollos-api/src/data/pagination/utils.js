// eslint-disable-next-line import/prefer-default-export
export const withEdgePagination = async ({ edges }) => {
  const result = await edges;
  const startCursor = result.length ? result[0].cursor : null;
  const endCursor = result.length ? result[result.length - 1].cursor : null;
  return { startCursor, endCursor };
};
