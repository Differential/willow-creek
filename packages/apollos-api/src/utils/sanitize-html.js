import sanitizeHtml from 'sanitize-html';

const allowedTags = [
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'blockquote',
  'p',
  'a',
  'ul',
  'ol',
  'li',
  'b',
  'i',
  'strong',
  'em',
  'br',
  'caption',
  'img',
];

const allowedAttributes = {
  a: ['href'],
  img: ['src'],
};

// A very picky HTML sanitizer
export default function(dirty) {
  return sanitizeHtml(dirty, {
    allowedTags,
    allowedAttributes,
  });
}
