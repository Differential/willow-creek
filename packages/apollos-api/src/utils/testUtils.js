import { KeyValueCache } from 'apollo-server-caching';
import {
  context as getContext,
  dataSources as getDataSources,
} from '../data/index';

export function getTestContext(req) {
  const context = getContext(req);
  const dataSources = getDataSources();
  // Apollo Server does this internally.
  Object.values(dataSources).forEach((dataSource) => {
    if (dataSource.initialize) {
      dataSource.initialize({ context, cache: KeyValueCache });
    }
  });
  context.dataSources = dataSources;
  return context;
}

export const buildGetMock = (response, dataSource) => {
  const get = jest.fn();
  if (Array.isArray(response) && Array.isArray(response[0])) {
    response.forEach((responseVal) => {
      get.mockReturnValueOnce(
        new Promise((resolve) => resolve(dataSource.normalize(responseVal)))
      );
    });
  }
  get.mockReturnValue(
    new Promise((resolve) => resolve(dataSource.normalize(response)))
  );
  return get;
};
