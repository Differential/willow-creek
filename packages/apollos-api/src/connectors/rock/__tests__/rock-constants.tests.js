import { fetch } from 'apollo-server-env';
import { buildGetMock } from '../../../utils/testUtils';

import RockConstants from '../rock-constants';

describe('RockConstants', () => {
  beforeEach(() => {
    fetch.resetMocks();
  });
  it("creates a Channel if it doesn't exist", async () => {
    const dataSource = new RockConstants();
    dataSource.get = buildGetMock([[], { Id: 1 }], dataSource);
    dataSource.post = buildGetMock('1', dataSource);
    const result = await dataSource.interactionChannel();
    expect(result).toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
    expect(dataSource.post.mock.calls).toMatchSnapshot();
  });
  it('finds the Channel if it exists', async () => {
    const dataSource = new RockConstants();
    dataSource.get = buildGetMock([{ Id: 1 }], dataSource);
    dataSource.post = jest.fn();
    const result = await dataSource.interactionChannel();
    expect(result).toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
    expect(dataSource.post.mock.calls.length).toBe(0);
  });
  it("creates a Component if it doesn't exist", async () => {
    const dataSource = new RockConstants();
    dataSource.get = buildGetMock([[], { Id: 1 }], dataSource);
    dataSource.post = buildGetMock('1', dataSource);
    const result = await dataSource.interactionComponent();
    expect(result).toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
    expect(dataSource.post.mock.calls).toMatchSnapshot();
  });
  it('finds the Component if it exists', async () => {
    const dataSource = new RockConstants();
    dataSource.get = buildGetMock([{ Id: 1 }], dataSource);
    dataSource.post = jest.fn();
    const result = await dataSource.interactionComponent();
    expect(result).toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
    expect(dataSource.post.mock.calls.length).toBe(0);
  });
  it('finds a ContentItem model ID', async () => {
    const dataSource = new RockConstants();
    dataSource.get = buildGetMock([{ Id: 1 }], dataSource);
    const result = await dataSource.modelTypeId('ContentItem');
    expect(result).toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
  });
  it('finds a UniversalContentItem model ID', async () => {
    const dataSource = new RockConstants();
    dataSource.get = buildGetMock([{ Id: 1 }], dataSource);
    const result = await dataSource.modelTypeId('UniversalContentItem');
    expect(result).toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
  });
  it('Throws when finding an unknown model ', () => {
    const dataSource = new RockConstants();
    dataSource.get = buildGetMock([{ Id: 1 }], dataSource);
    const prom = dataSource.modelTypeId('IDontExist');
    expect(prom).rejects.toEqual(
      new Error('IDontExist has not been mapped into a Rock type!')
    );
  });
});
