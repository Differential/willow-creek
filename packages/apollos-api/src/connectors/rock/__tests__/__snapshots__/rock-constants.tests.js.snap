// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RockConstants creates a Channel if it doesn't exist 1`] = `
Object {
  "id": 1,
}
`;

exports[`RockConstants creates a Channel if it doesn't exist 2`] = `
Array [
  Array [
    "InteractionChannels?%24filter=%28Name%20eq%20%27Apollos%20App%27%29%20and%20%28UsesSession%20eq%20true%29",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
  Array [
    "/InteractionChannels/1",
  ],
]
`;

exports[`RockConstants creates a Channel if it doesn't exist 3`] = `
Array [
  Array [
    "/InteractionChannels",
    Object {
      "Name": "Apollos App",
      "UsesSession": true,
    },
  ],
]
`;

exports[`RockConstants creates a Component if it doesn't exist 1`] = `Array []`;

exports[`RockConstants creates a Component if it doesn't exist 2`] = `
Array [
  Array [
    "InteractionChannels?%24filter=%28Name%20eq%20%27Apollos%20App%27%29%20and%20%28UsesSession%20eq%20true%29",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
  Array [
    "/InteractionChannels/1",
  ],
  Array [
    "InteractionComponents?%24filter=%28Name%20eq%20%27Apollos%20Content%20Item%27%29%20and%20%28ChannelId%20eq%201%29",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
]
`;

exports[`RockConstants creates a Component if it doesn't exist 3`] = `
Array [
  Array [
    "/InteractionChannels",
    Object {
      "Name": "Apollos App",
      "UsesSession": true,
    },
  ],
]
`;

exports[`RockConstants finds a ContentItem model ID 1`] = `
Object {
  "id": 1,
}
`;

exports[`RockConstants finds a ContentItem model ID 2`] = `
Array [
  Array [
    "EntityTypes?%24filter=Name%20eq%20%27Rock.Model.ContentChannelItem%27",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
]
`;

exports[`RockConstants finds a UniversalContentItem model ID 1`] = `
Object {
  "id": 1,
}
`;

exports[`RockConstants finds a UniversalContentItem model ID 2`] = `
Array [
  Array [
    "EntityTypes?%24filter=Name%20eq%20%27Rock.Model.ContentChannelItem%27",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
]
`;

exports[`RockConstants finds the Channel if it exists 1`] = `
Object {
  "id": 1,
}
`;

exports[`RockConstants finds the Channel if it exists 2`] = `
Array [
  Array [
    "InteractionChannels?%24filter=%28Name%20eq%20%27Apollos%20App%27%29%20and%20%28UsesSession%20eq%20true%29",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
]
`;

exports[`RockConstants finds the Component if it exists 1`] = `
Object {
  "id": 1,
}
`;

exports[`RockConstants finds the Component if it exists 2`] = `
Array [
  Array [
    "InteractionChannels?%24filter=%28Name%20eq%20%27Apollos%20App%27%29%20and%20%28UsesSession%20eq%20true%29",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
  Array [
    "InteractionComponents?%24filter=%28Name%20eq%20%27Apollos%20Content%20Item%27%29%20and%20%28ChannelId%20eq%201%29",
    Object {},
    Object {
      "ttl": 86400,
    },
  ],
]
`;
